src/Algebra/Ordered/Structures.agda

```diff
@@ -0,0 +1,205 @@
+      identity       : Identity _≈_ ε _∙_
+
+    open IsProsemigroup isProsemigroup public
+
+    isMonoid : IsMonoid _≈_ _∙_ ε
+    isMonoid = record
+      { isSemigroup = isSemigroup
+      ; identity    = identity
+      }
+
+    open IsMonoid isMonoid public using (identityˡ; identityʳ)
+
+  record IsCommutativePromonoid (_∙_ : Op₂ A) (ε : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isPromonoid : IsPromonoid _∙_ ε
+      comm        : Commutative _≈_ _∙_
+
+    open IsPromonoid isPromonoid public
+
+    isCommutativeMonoid : IsCommutativeMonoid _≈_ _∙_ ε
+    isCommutativeMonoid = record
+      { isMonoid        = isMonoid
+      ; comm            = comm
+      }
+
+  -- Preordered semirings (prosemirings)
+
+  record IsProsemiring (_+_ _*_ : Op₂ A) (0# 1# : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      +-isCommutativePromonoid : IsCommutativePromonoid _+_ 0#
+      *-isPromonoid            : IsPromonoid _*_ 1#
+      distrib                  : _DistributesOver_ _≈_ _*_ _+_
+      zero                     : Zero _≈_ 0# _*_
+
+    distribˡ : _DistributesOverˡ_ _≈_ _*_ _+_
+    distribˡ = proj₁ distrib
+
+    distribʳ : _DistributesOverʳ_ _≈_ _*_ _+_
+    distribʳ = proj₂ distrib
+
+    zeroˡ : LeftZero _≈_ 0# _*_
+    zeroˡ = proj₁ zero
+
+    zeroʳ : RightZero _≈_ 0# _*_
+    zeroʳ = proj₂ zero
+
+    open IsCommutativePromonoid +-isCommutativePromonoid public
+      renaming
+      ( assoc       to +-assoc
+      ; ∙-cong      to +-cong
+      ; monotonic   to +-monotonic
+      ; identity    to +-identity
+      ; identityˡ   to +-identityˡ
+      ; identityʳ   to +-identityʳ
+      ; comm        to +-comm
+      ; isMagma     to +-isMagma
+      ; isSemigroup to +-isSemigroup
+      ; isMonoid    to +-isMonoid
+      )
+
+    open IsPromonoid *-isPromonoid public
+      using ()
+      renaming
+      ( assoc       to *-assoc
+      ; ∙-cong      to *-cong
+      ; monotonic   to *-monotonic
+      ; identity    to *-identity
+      ; identityˡ   to *-identityˡ
+      ; identityʳ   to *-identityʳ
+      ; isMagma     to *-isMagma
+      ; isSemigroup to *-isSemigroup
+      ; isMonoid    to *-isMonoid
+      )
+
+------------------------------------------------------------------------
+-- Partially ordered structures
+
+module _ {a ℓ₁ ℓ₂} {A : Set a}
+  (_≈_ : Rel A ℓ₁)  -- The underlying equality.
+  (_≤_ : Rel A ℓ₂)  -- The partial.
+  where
+
+  -- Partially ordered semigroups (posemigroups)
+
+  record IsPosemigroup (_∙_ : Op₂ A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isPartialOrder : IsPartialOrder _≈_ _≤_
+      assoc          : Associative _≈_ _∙_
+      monotonic      : Monotonic₂ _≤_ _∙_
+
+    open IsPartialOrder isPartialOrder public
+
+    ∙-cong : Congruent₂ _≈_ _∙_
+    ∙-cong x≈y u≈v =
+      antisym (monotonic (reflexive x≈y) (reflexive u≈v))
+              (monotonic (reflexive (Eq.sym x≈y)) (reflexive (Eq.sym u≈v)))
+
+    isProsemigroup : IsProsemigroup _≈_ _≤_ _∙_
+    isProsemigroup = record
```

Comment:
Case in point. If you fix the lack of symmetry, then you won't have to build this entire structure right?

Reply:
Indeed, but see my earlier comment about redundant fields...